lib/auth/methods.go

```diff
@@ -0,0 +1,428 @@
+	Session *SessionCreds `json:"session,omitempty"`
+}
+
+// CheckAndSetDefaults checks and sets defaults
+func (a *AuthenticateUserRequest) CheckAndSetDefaults() error {
+	if a.Username == "" {
+		return trace.BadParameter("missing parameter 'username'")
+	}
+	if a.Pass == nil && a.U2F == nil && a.OTP == nil && a.Session == nil {
+		return trace.BadParameter("at least one authentication method is required")
+	}
+	return nil
+}
+
+// PassCreds is a password credential
+type PassCreds struct {
+	// Password is a user password
+	Password []byte `json:"password"`
+}
+
+// U2FSignResponseCreds is a U2F signature sent by U2F device
+type U2FSignResponseCreds struct {
+	// SignResponse is a U2F sign resposne
+	SignResponse u2f.SignResponse `json:"sign_response"`
+}
+
+// OTPCreds is a two factor authencication credentials
+type OTPCreds struct {
+	// Password is a user password
+	Password []byte `json:"password"`
+	// Token is a user second factor token
+	Token string `json:"token"`
+}
+
+// SessionCreds is a web session credentials
+type SessionCreds struct {
+	// ID is a web session id
+	ID string `json:"id"`
+}
+
+// AuthenticateUser authenticates user based on the request type
+func (s *AuthServer) AuthenticateUser(req AuthenticateUserRequest) error {
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	authPreference, err := s.GetAuthPreference()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	switch {
+	case req.Pass != nil:
+		// authenticate using password only, make sure
+		// that auth preference does not require second factor
+		// otherwise users can bypass the second factor
+		if authPreference.GetSecondFactor() != teleport.OFF {
+			return trace.AccessDenied("missing second factor")
+		}
+		err := s.WithUserLock(req.Username, func() error {
+			return s.CheckPasswordWOToken(req.Username, req.Pass.Password)
+		})
+		if err != nil {
+			// provide obscure message on purpose, while logging the real
+			// error server side
+			log.Debugf("failed to authenticate: %v", err)
+			return trace.AccessDenied("invalid username or password")
+		}
+		return nil
+	case req.U2F != nil:
+		// authenticate using U2F - code checks challenge response
+		// signed by U2F device of the user
+		err := s.WithUserLock(req.Username, func() error {
+			return s.CheckU2FSignResponse(req.Username, &req.U2F.SignResponse)
+		})
+		if err != nil {
+			// provide obscure message on purpose, while logging the real
+			// error server side
+			log.Debugf("failed to authenticate: %v", err)
+			return trace.AccessDenied("invalid U2F response")
+		}
+		return nil
+	case req.OTP != nil:
+		err := s.WithUserLock(req.Username, func() error {
+			return s.CheckPassword(req.Username, req.OTP.Password, req.OTP.Token)
+		})
+		if err != nil {
+			// provide obscure message on purpose, while logging the real
+			// error server side
+			log.Debugf("failed to authenticate: %v", err)
```

Comment:
"Failed", dot at the end, trace.DebugReport. You could probably move the error handling outside the switch as all methods handle it same way.